### PR TITLE
SONAR-7254 Suppress "sonar.security.localUsers" property

### PR DESCRIPTION
**Pull request by julienlancelot**

### CODE REVIEW

File: sonar-plugin-api/src/main/java/org/sonar/api/CoreProperties.java [hunk 1]
```diff
@@ -468,7 +468,9 @@
 
   /**
    * @since 4.2
+   * @deprecated no more used since 5.5
    */
+  @Deprecated
   String CORE_AUTHENTICATOR_LOCAL_USERS = "sonar.security.localUsers";
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/CoreProperties.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)

File: sonar-db/src/main/java/org/sonar/db/user/UserDao.java [hunk 2]
```diff
@@ -177,11 +198,27 @@ public UserDto selectOrFailByLogin(DbSession session, String login) {
...
+
+    private LoginToUser(Collection<UserDto> unordered) {
+      for (UserDto dto : unordered) {
+        map.put(dto.getLogin(), dto);
+      }
+    }
+
+    @Nullable
+    @Override
+    public UserDto apply(@Nonnull String login) {
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/user/UserDao.java]: MAJOR Remove this "Nonnull" annotation to honor the overridden method's contract. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2638)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 4 major
* INFO 1 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Remove this unused "userIndex" private field.](https://github.com/SonarSource/sonarqube/blob/79ab51af24bfd7ed2cb0ba01387695352e0bbbd7/server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CreateAction.java#L59) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1068)
* MAJOR [Remove this unused "userIndex" private field.](https://github.com/SonarSource/sonarqube/blob/79ab51af24bfd7ed2cb0ba01387695352e0bbbd7/server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/SearchAction.java#L61) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1068)
* MAJOR [Remove this unused "userIndex" private field.](https://github.com/SonarSource/sonarqube/blob/79ab51af24bfd7ed2cb0ba01387695352e0bbbd7/server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/UpdateAction.java#L52) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1068)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 1 major
* INFO 1 info

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 1 major
* INFO 1 info

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/startup/FeedUsersLocalStartupTask.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+import static org.sonar.db.loadedtemplate.LoadedTemplateDto.ONE_SHOT_TASK_TYPE;
+
+/**
+ * Feed users local property.
+ * If a realm is defined, then users are set as local only if their login are found in the property "sonar.security.localUsers",
+ * otherwise user are all set as local.
+ *
+ * See <a href="https://jira.sonarsource.com/browse/SONAR-7254">SONAR-7254</a>.
+ *
+ * Should be removed after LTS 5.X
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/startup/FeedUsersLocalStartupTask.java]: can you create a jira ticket if it's not already done?

File: server/sonar-server/src/main/java/org/sonar/server/startup/FeedUsersLocalStartupTask.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+import org.sonar.db.user.UserDto;
+import org.sonar.server.user.UserUpdater;
+
+import static java.util.Arrays.asList;
+import static org.sonar.api.CoreProperties.CORE_AUTHENTICATOR_REALM;
+import static org.sonar.db.loadedtemplate.LoadedTemplateDto.ONE_SHOT_TASK_TYPE;
+
+/**
+ * Feed users local property.
+ * If a realm is defined, then users are set as local only if their login are found in the property "sonar.security.localUsers",
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/startup/FeedUsersLocalStartupTask.java]: you confirm it's already the behavior? So when migrating, if users created in SQ but not listed in sonar.security.localUsers will be considered non local?

File: server/sonar-server/src/main/java/org/sonar/server/startup/FeedUsersLocalStartupTask.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+  @Override
+  public void start() {
+    DbSession session = dbClient.openSession(false);
+    try {
+      if (hasAlreadyBeenExecuted(session)) {
+        return;
+      }
+      updateUsersLocal(session);
+      markAsExecuted(session);
+      session.commit();
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/startup/FeedUsersLocalStartupTask.java]: session -> dbSession

File: server/sonar-server/src/main/java/org/sonar/server/startup/FeedUsersLocalStartupTask.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    } finally {
+      dbClient.closeSession(session);
+    }
+  }
+
+  private void updateUsersLocal(DbSession session) {
+    long now = system2.now();
+    Set<String> localUsers = new HashSet<>(asList(settings.getStringArray(LOCAL_USERS_PROPERTY)));
+    boolean isRealmExist = settings.getString(CORE_AUTHENTICATOR_REALM) != null;
+    for (UserDto user : dbClient.userDao().selectUsers(session, UserQuery.ALL_ACTIVES)) {
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/startup/FeedUsersLocalStartupTask.java]: we don't need to update non active users? They can be reactivated, that's why I'm wondering

File: server/sonar-server/src/main/java/org/sonar/server/user/ws/DeactivateAction.java [hunk 7]
```diff
@@ -83,21 +80,21 @@ public void handle(Request request, Response response) throws Exception {
...
   }
 
-  private void writeUser(JsonWriter json, UserDoc user) {
+  private void writeUser(JsonWriter json, String login) {
     json.name("user");
     Set<String> groups = Sets.newHashSet();
     DbSession dbSession = dbClient.openSession(false);
     try {
-      groups.addAll(dbClient.groupMembershipDao().selectGroupsByLogins(dbSession, Arrays.asList(user.login())).get(user.login()));
+      UserDto user = dbClient.userDao().selectOrFailByLogin(dbSession, login);
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/user/ws/DeactivateAction.java]: here `dbClient.userDao().selectOrFailByLogin(dbSession, login)` shouldn't be used. The dao layer returns a 500 HTTP error whereas we want a 401 in this specific case, as it's a user input. Si ComponentFinder for example, that return server exception

File: server/sonar-server/src/main/java/org/sonar/server/user/ws/UpdateAction.java [hunk 8]
```diff
@@ -110,21 +106,21 @@ public void handle(Request request, Response response) throws Exception {
...
   }
 
-  private void writeUser(JsonWriter json, UserDoc user) {
+  private void writeUser(JsonWriter json, String login) {
     json.name("user");
     Set<String> groups = Sets.newHashSet();
     DbSession dbSession = dbClient.openSession(false);
     try {
-      groups.addAll(dbClient.groupMembershipDao().selectGroupsByLogins(dbSession, Arrays.asList(user.login())).get(user.login()));
+      UserDto user = dbClient.userDao().selectOrFailByLogin(dbSession, login);
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/user/ws/UpdateAction.java]: `dbClient.userDao().selectOrFailByLogin(dbSession, login)` shouldn't be used, throws HTTP 500 error

File: server/sonar-server/src/main/java/org/sonar/server/startup/FeedUsersLocalStartupTask.java (hunk 3 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/startup/FeedUsersLocalStartupTask.java]: I think we should simply do some big clean up when we'll be in 6.0

File: server/sonar-server/src/main/java/org/sonar/server/startup/FeedUsersLocalStartupTask.java (hunk 4 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/startup/FeedUsersLocalStartupTask.java]: Have you tested it :) ?

File: server/sonar-server/src/main/java/org/sonar/server/startup/FeedUsersLocalStartupTask.java (hunk 5 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/startup/FeedUsersLocalStartupTask.java]: I'm not sure it really matters, does it ?

File: server/sonar-server/src/main/java/org/sonar/server/startup/FeedUsersLocalStartupTask.java (hunk 6 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/startup/FeedUsersLocalStartupTask.java]: Non active users will be updated when being reactivated

File: server/sonar-server/src/main/java/org/sonar/server/user/ws/DeactivateAction.java (hunk 7 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/user/ws/DeactivateAction.java]: You're right

File: server/sonar-server/src/main/java/org/sonar/server/startup/FeedUsersLocalStartupTask.java (hunk 4 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/startup/FeedUsersLocalStartupTask.java]: More seriously, yes, the JavaDoc is explaining what is happening.

File: server/sonar-server/src/main/java/org/sonar/server/user/ws/UpdateAction.java (hunk 8 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/user/ws/UpdateAction.java]: right

**Comment by julienlancelot** [comment]
Thanks for your feedback @teryk ! Please have a look at my last commit.

File: server/sonar-server/src/test/java/org/sonar/server/user/ws/DeactivateActionTest.java [hunk 9]
```diff
@@ -124,7 +124,7 @@ public void fail_on_missing_permission() throws Exception {
       .setParam("login", "john").execute();
   }
 
-  @Test(expected = RowNotFoundException.class)
+  @Test(expected = NotFoundException.class)
```
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/user/ws/DeactivateActionTest.java]: I'd prefer to use the `ExpectedException` rule but very minor

**Comment by teryk** [comment]
lgtm-jesus

File: server/sonar-server/src/test/java/org/sonar/server/user/ws/DeactivateActionTest.java (hunk 9 above)
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/user/ws/DeactivateActionTest.java]: Same for me, but I was too lazy this time :(

File: sonar-db/src/main/java/org/sonar/db/user/UserDao.java [hunk 10]
```diff
@@ -177,11 +197,26 @@ public UserDto selectOrFailByLogin(DbSession session, String login) {
...
+    private final Map<String, UserDto> map = new HashMap<>();
+
+    private LoginToUser(Collection<UserDto> unordered) {
+      for (UserDto dto : unordered) {
+        map.put(dto.getLogin(), dto);
+      }
+    }
+
+    @Override
+    public UserDto apply(@Nonnull String login) {
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/user/UserDao.java]: MAJOR Remove this "Nonnull" annotation to honor the overridden method's contract. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2638)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 1 major
* INFO 1 info

Watch the comments in this conversation to review them.

**Pull request closed without merging**